airbyte-integrations/connectors/source-netsuite/source_netsuite/streams.py

```diff
@@ -0,0 +1,266 @@
+#
+# Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+#
+
+
+from abc import ABC
+from datetime import date, datetime, timedelta
+from typing import Any, Iterable, Mapping, MutableMapping, Optional, Union
+
+import requests
+from airbyte_cdk.sources.streams.http import HttpStream
+from requests_oauthlib import OAuth1
+
+# paths for NetSuite REST API
+REST_PATH: str = "/services/rest/"
+RECORD_PATH: str = REST_PATH + "record/v1/"
+META_PATH: str = RECORD_PATH + "metadata-catalog/"
+# predefine header
+SCHEMA_HEADERS: dict = {"Accept": "application/schema+json"}
+# known error codes by their HTTP codes
+NETSUITE_ERRORS_MAPPING: dict = {
+    400: {
+        "USER_ERROR": "reading an Admin record allowed for Admin only",
+        "NONEXISTENT_FIELD": "cursor_field declared in schema but doesn't exist in object",
+        "INVALID_PARAMETER": "cannot read or find the object. Skipping",
+    },
+}
+
+
+class NetsuiteStream(HttpStream, ABC):
+    def __init__(
+        self,
+        auth: OAuth1,
+        obj_name: str,
+        base_url: str,
+        start_datetime: str,
+        window_in_days: int,
+    ):
+        self.obj_name = obj_name
+        self.base_url = base_url
+        self.start_datetime = start_datetime
+        self.window_in_days = window_in_days
+        self.schemas = {}  # store subschemas to reduce API calls
+        super().__init__(authenticator=auth)
+
+    primary_key = "id"
+    raise_on_http_errors = True
+    output_datetime_format = "%Y-%m-%dT%H:%M:%SZ"
+    input_datetime_format = "%m/%d/%Y"
+
+    @property
+    def name(self) -> str:
+        return self.obj_name
+
+    @property
+    def url_base(self) -> str:
+        return self.base_url
+
+    def path(self, **kwargs) -> str:
+        return RECORD_PATH + self.obj_name
+
+    def ref_schema(self) -> Mapping[str, str]:
+        return {
+            "type": ["null", "object"],
+            "properties": {
+                "id": {"title": "Internal identifier", "type": ["null", "string"]},
```

Comment:
null values shouldn't be allowed as long as id is the primary key

Reply:
This is the nested object that contains links, this is not PK.